Implement predict_quantiles and predict_intervals in theta forecasters, tests in base class

kejsitake opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look
at our contribution guide: https://github.com/alan-turing-institute/sktime/blob/main/CONTRIBUTING.md
-->

#### Reference Issues/PRs
<!--
Example: Fixes #1234. See also #3456.

Please use keywords (e.g., Fixes) to create link to the issues or pull requests
you resolved, so that they will automatically be closed when your pull request
is merged. See https://github.com/blog/1506-closing-issues-via-pull-requests
-->
Related to https://github.com/alan-turing-institute/sktime/issues/984

#### What does this implement/fix? Explain your changes.
<!--
A clear and concise description of what you have implemented.
-->
Implements a function to test predict_quantiles
#### Does your contribution introduce a new dependency? If yes, which one?

<!--
If your contribution does add a new hard dependency, we may suggest to initially develop your contribution in a separate companion package in https://github.com/sktime/ to keep external dependencies of the core sktime package to a minimum.
-->
No
#### What should a reviewer concentrate their feedback on?

<!-- This section is particularly useful if you have a pull request that is still in development. You can guide the reviews to focus on the parts that are ready for their comments. We suggest using bullets (indicated by * or -) and filled checkboxes [x] here -->
Implemented these checks:
- [X] check if the multi-indexed column names are the same as the expected ones
- [X] check if the row indexes or the time frame are the right ones 
- [X] check that the predicted quantiles are strictly monotonically increasing

Modified extension docs to account for base forecasters class:
- [X] removed _compute_pred_int because according to the STEP 13 that should only be in theta
- [X] added _predict_interval function
- [X] added _predict_quantiles
- [X] added capability:pred_int to the tags

Changes in the Theta classifier:
- [X] Implemented predict_quantiles using compute_pred_int
- [X] Implemented predict_interval

#### Any other comments?
<!--
Please be aware that we are a loose team of volunteers so patience is necessary; assistance handling other issues is very welcome. We value all user contributions, no matter how minor they are. If we are slow to review, either the pull request needs some benchmarking, tinkering, convincing, etc. or more likely the reviewers are simply busy. In either case, we ask for your understanding during the review process.
-->

#### PR checklist
<!--
Please go through the checklist below. Please feel free to remove points if they are not applicable.
-->

<!--
Thanks for contributing!
-->

fkiraly commented on sktime/forecasting/tests/test_all_forecasters.py:
```diff
@@ -286,11 +287,56 @@ def test_predict_pred_interval(Forecaster, fh, alpha):
+    # alpha can be a float or a list
+    # if a list, it has to be sorted
+    # check if alpha is increasing and in the given interval
+    if isinstance(alpha, list):
```
hm, this looks more like a test that the `BaseForecaster` should carry out? `alpha` is an input, and we know that the `alpha` in the test are fine? 

I suggest:
* move this to the `BaseForecaster` as an input check for the `alpha`; extend this that it checks that `alpha` is a `float`, or list of `float`.
* add a test which checks that `BaseForecaster` raises an error for "wrong" `alpha`.

fkiraly replied:
isn't there already a `check_alpha`, btw?

fkiraly commented on sktime/forecasting/tests/test_all_forecasters.py:
```diff
@@ -286,11 +287,56 @@ def test_predict_pred_interval(Forecaster, fh, alpha):
+    # check if values are monotonically increasing
+    for var in pred_quantiles.columns.levels[0]:
+        for index in range(len(pred_quantiles.index)):
+            assert pred_quantiles[var].iloc[index].is_monotonic_increasing
```
this doesn't always give the expected result, since the user could pass `alpha` in the list in a non-monotonic order? An easy way around this is to retrieve the alpha, and sort the columns before checking.

fkiraly commented:
@kejsitake, it seems your failures are related to linting, i.e., code formatting conventions that `sktime` enforces.
For instance, use of double-quote instead of quote in strings.

You can check what is going wrong by clicking on "details" next to the failing test, or locally in your dev IDE if you have linting and testing set up there. Let us know in the community collab session if your local set-up is not fully working, this can be fiddly, especially on Windows.

kejsitake commented:
I finally figured out how to run pre-commit locally. However, the PR is not ready for review yet, just commiting my changes.

fkiraly commented on extension_templates/forecasting.py:
```diff
@@ -90,6 +90,7 @@ class MyForecaster(BaseForecaster):
         "requires-fh-in-fit": True,  # is forecasting horizon already required in fit?
         "X-y-must-have-same-index": True,  # can estimator handle different X/y index?
         "enforce_index_type": None,  # index type that needs to be enforced in X/y
+        "capability:pred_int": False,  # is forecaster probabilistic?
```
maybe a better description: "does forecaster implement `predict_quantiles`?"

Also consider renaming this to `"capability:predict_quantiles"`, but perhaps it's best to leave it like this since people may know the tag.

fkiraly commented on extension_templates/forecasting.py:
```diff
@@ -238,19 +239,62 @@ def _update_predict_single(
     # if not implementing, delete the method
-    def _compute_pred_int(self, alphas):
-        """Calculate the prediction errors for each point.
+    def _predict_interval(self, fh, X, coverage):
```
right idea, although I think it may be confusing. In the extension template I'd only include `_predict_quantiles`, because `_predict_interval` automatically calls that. I'd add an explanation that the user can implement `_predict_interval` as well if there is a faster variant, but I think it is so rare that it shouldn't be in the template as a "fill in" function, but only mentioned in the body of `_predict_quantiles`.

fkiraly commented on extension_templates/forecasting.py:
```diff
@@ -238,19 +239,62 @@ def _update_predict_single(
 
     # todo: consider implementing this, optional
     # if not implementing, delete the method
-    def _compute_pred_int(self, alphas):
```
important: we also need to remove the "old" prediction interval related arguments from `_predict`, and similarly from the docstring preamble at the top (`return_pred_int` etc)

fkiraly commented on sktime/forecasting/tests/test_all_forecasters.py:
```diff
@@ -278,19 +279,61 @@ def test_predict_pred_interval(Forecaster, fh, alpha):
         else:
-            with pytest.raises(NotImplementedError, match="prediction intervals"):
-                f.predict(return_pred_int=True, alpha=alpha)
+            f.predict(alpha=alpha)
```
I'm confused about this line - should this not be `f.predict_interval`?

fkiraly replied:
we also need to do something to check the return. It should be in the expected format.

fkiraly commented on sktime/forecasting/tests/test_all_forecasters.py:
```diff
@@ -278,19 +279,61 @@ def test_predict_pred_interval(Forecaster, fh, alpha):
+            f.predict(alpha=alpha)
+
+
+def _check_predict_quantiles(pred_quantiles: list, y_train: pd.Series, fh, alpha):
```
we should probably also test this for multivariate `y`, too

fkiraly replied:
you do this below, but I think the type hints are inaccurate

fkiraly commented on sktime/forecasting/theta.py:
```diff
@@ -162,9 +162,9 @@ def _predict(self, fh, X=None, return_pred_int=False, alpha=DEFAULT_ALPHA):
-        if return_pred_int:
-            pred_int = self.compute_pred_int(y_pred=y_pred, alpha=alpha)
-            return y_pred, pred_int
+        # if return_pred_int:
```
the `return_pred_int` and `alpha` arguments should also be removed

fkiraly reviewed (changes requested):
Thanks, nice work!

First of all, I'd suggest to move the work on the extension template into a separate PR if it's not too much of a hassle. This is independent, so could be independently reviewed on. I feel we're almost there with that.

On the tests:
* tests look good in general!
* I think there's a problem with the logic around handling whether the method has been refactored already or not. I think methods that haven't been refactored will currently break if their `predict_interval` or `predict_quantiles` are run? The tests don't account for that (though I may be mistaken). Perhaps have another look at what happens for refactored, and non-refactored estimators?

On theta:
* I think this is very convoluted - methods calling methods calling methods unnevessarily, i.e., `predict_quantiles` calls `compute_pred_int` which calls `_compute_pred_err`. I think out of the three items, there is still most work needed. What I suggest below will involve deleting a lot of lines of code, so the work is in carefully thinking what calls what and cutting down unnecessary layers.
* Suggested strategy: just kill `compute_pred_int` and `_compute_pred_err` entirely, and move the real logic into `predict_quantiles`. It will be much nicer (and less convoluted).
* `_predict_interval` is not needed at all, once you've implemented `predict_quantiles`.

kejsitake commented:
Thank you for the comments @fkiraly.

I believe I have solved most of them. I noticed that one of the tests fails. Upon looking more at the issue, this is happening because there is still a test `test_predict_pred_interval` for the old format of `pred_ints`, which expects the columns of the returned dataframe to be `lower` and `upper`.  But, the new predict_interval calls `predict_quantiles` and returns a dataframe in the new format(with the multi index), causing the test to fail. I am not sure if it is safe to remove this test, or we should keep it at this time, and I should instead look into convert the format to the old one.

kejsitake commented:
I thought some more about this. In my opinion the test for the old format of the output can be removed. I wasn't sure about the format that predict_intervals should return. Currently it returns the same format dataframe as predict_quantiles. I thought about renaming the first level to "Intervals" instead of "Quantiles", but that didn't seem correct either. I didn't find a specification about this on the STEP either, but please let me know if I am missing anything.

fkiraly commented:
> But, the new predict_interval calls predict_quantiles and returns a dataframe in the new format(with the multi index), causing the test to fail. I am not sure if it is safe to remove this test, or we should keep it at this time, and I should instead look into convert the format to the old one.

The story is, `predict` should still return the *old* format correctly until we have deprecated it completely.
For this, the function `_convert_new_to_old_pred_int` has been written, it should convert the new format to the old format.

So, the test should pass if `_predict_quantiles` is implemented correctly and `_convert_new_to_old_pred_int` is correct, which may not be the case (I would recommend to theck).

Due to the deprecation, we shouldn't remove the "old" test yet - we should remove it when we deprecate the old prediction interval interface.

fkiraly reviewed (changes requested):
Yes, the earlier issues seem to have been resolved.

Major comments:
* you accidentally deleted the `forecaster` extension template? This should be reverted.
* the test for the old prediction intervals should pass, it should work out-of-the box via `_convert_new_to_old_pred_int`. Could you kindly check why it doesn't? Could be the implementation, could be the conversion. May be worth asking @SveaMeyer13 who I believe wrote the conversion, if the test failure can be localized to there. But it could also be that you are returning the wrong format in `_predict_quantiles`, or sth with the base class.

kejsitake commented:
Just summarizing my latest changes on this PR to make it easier to review:

- I reverted the forecaster extension template, which I had removed by mistake when splitting this PR.
- I added the conversion to the old format and the old predict interval tests now pass
- I fixed a bug related to rounding of c during division by converting it to a float.

All the tests seem to pass successfully now.

fkiraly reviewed (dismissed):
Looks good, issues seem to have been addressed.

Could be merged if it weren't up to the merge conflict which should be resolved.

fkiraly commented on sktime/forecasting/tests/test_theta.py:
```diff
@@ -1,4 +1,9 @@
+# copyright: sktime developers, BSD-3-Clause License (see LICENSE file)
+"""
+
 __author__ = ["@big-o"]
```
kindly add yourself, too to this

fkiraly commented:
also, could you kindly comment on the open discussions, how they have been resolved and click "reolve"? So unresolved conversations = 0.

fkiraly reviewed (approved):
looks good now, thanks!
